### PR TITLE
add the LinearChainCrf operator.

### PR DESCRIPTION
**Pull request by lcy-seso**
fixes https://github.com/PaddlePaddle/Paddle/issues/4168
fixes https://github.com/PaddlePaddle/Paddle/issues/5250

**Currently, this operator only runs on CPU.**

1. This operator has been manually tested with the old CRFLayer. Their computation results have no differences.
~~2. Currently, the linear chain CRF operator does not backpropagate the gradients it receives, so it can only be used at the end of the network. This can be easily fixed.~~
3. The following things still need to be optimized:
    ~~- It is better to use double in unitest for this operator. But, currently, I come into a coredump and I haven't checked it yet. I will fix this later.~~
    - It is better if some local intermediate variable is optimized to avoid frequently allocating/freeing memory.

### CODE REVIEW

File: paddle/operators/linear_chain_crf_op.cc [hunk 1]
```diff
@@ -0,0 +1,567 @@
...
+        }
+        beta_value[k * tag_num + i] = sum;
+      }
+      NormalizeL1<T>(beta_value + k * tag_num, tag_num);
+    }
+
+    auto alpha_mat = EigenMatrix<T>::From(*alpha);
+    auto beta_mat = EigenMatrix<T>::From(*beta);
+    auto x_grad_mat = EigenMatrix<T>::From(*emission_grad);
+    auto* place = ctx.GetEigenDevice<platform::CPUPlace>();
```
Comment (pengli09) [inline paddle/operators/linear_chain_crf_op.cc]: change `auto*` to `auto`

File: paddle/operators/linear_chain_crf_op.cc [hunk 2]
```diff
@@ -0,0 +1,567 @@
...
+    auto prob = alpha_mat * beta_mat;
+    auto row_sum = prob.sum(Eigen::DSizes<int, 1>(1))
+                       .reshape(Eigen::DSizes<int, 2>(seq_length, 1))
+                       .broadcast(Eigen::DSizes<int, 2>(1, tag_num));
+    x_grad_mat.device(*place) = prob / row_sum;
```
Comment (pengli09) [inline paddle/operators/linear_chain_crf_op.cc]: change `*place` to `place`

File: python/paddle/v2/framework/tests/test_linear_chain_crf_op.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+class TestLinearChainCrfOp(OpTest):
+    def set_test_data(self):
+        SEQ_NUM = 3
+        TAG_NUM = 17
+        MAX_SEQ_LEN = 5
+
+        # the linear_chain_crf operator only supports sequence (LoD level = 1)
+        lod = [[0]]
+        for i in range(SEQ_NUM):
+            lod[-1].append(lod[-1][-1] + random.randint(1, MAX_SEQ_LEN))
```
Comment (pengli09) [inline python/paddle/v2/framework/tests/test_linear_chain_crf_op.py]: Sequence of length 1, 2 and 3 must be included in the test. These are high-risk boundary conditions.

File: python/paddle/v2/framework/tests/test_linear_chain_crf_op.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
+import unittest
+import random
+import numpy as np
+
+from op_test import OpTest
+
+
+class LinearChainCrfForward(object):
```
Comment (pengli09) [inline python/paddle/v2/framework/tests/test_linear_chain_crf_op.py]: A test case of LinearChainCrfForward() should also be included to make sure the implementation is correct. Furthermore, it is also useful for detecting unexpected errors introduced by future modifications. Reference: https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/gserver/tests/test_CRFLayerGrad.cpp#L56

File: paddle/operators/linear_chain_crf_op.cc [hunk 5]
```diff
@@ -0,0 +1,567 @@
...
+        "(LoDTensor, default: LoDTensor<int>). The ground truth which is a 2-D "
+        "LoDTensor with shape [N x 1], where N is the total element number in "
+        "a mini-batch.");
+    AddOutput(
+        "Alpha",
+        "Tensor, default: Tensor<float>. The forward vectors for the entire "
+        "batch. A two dimensional tensor with shape [N x D], "
+        "denoted as \f$\alpha\f$. \f$\alpha$\f is a memo table used to "
+        "calculate the normalization factor in CRF. \f$\alpha[k, v]$\f stores "
+        "the unnormalized probabilites of all possible unfinished sequences of "
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply paddle/operators/linear_chain_crf_op.cc]: The "unnormalized probabilities" is just from the paper.

File: paddle/operators/linear_chain_crf_op.cc [hunk 6]
```diff
@@ -0,0 +1,570 @@
...
+
+    ctx->SetOutputDim("Alpha", emission_dims);
+    ctx->SetOutputDim("EmissionExps", emission_dims);
+    ctx->SetOutputDim("TransitionExps", transition_dims);
+    // (TODO caoying) This is tricky. The 1st dimension of Output(LogLikelihood)
+    // is the sequence number in a mini-batch. The dimension set here should be
+    // resized to its correct size in the function Compute.
+    ctx->SetOutputDim("LogLikelihood", {emission_dims[0], 1});
+
+    ctx->ShareLoD("Emission", /*->*/ "EmissionExps");
```
Comment (qingqing01) [inline paddle/operators/linear_chain_crf_op.cc]: If the output is not a LoDTensor, this line can be removed.

File: paddle/operators/linear_chain_crf_op.cc [hunk 7]
```diff
@@ -0,0 +1,570 @@
...
+                   "Input(EmissionExps) should be not null.");
+    PADDLE_ENFORCE(ctx->HasInput("TransitionExps"),
+                   "Input(TransitionExps) should be not null.");
+    PADDLE_ENFORCE(ctx->HasInput(framework::GradVarName("LogLikelihood")),
+                   "Input(LogLikelihood@GRAD) shoudl be not null.");
+
+    PADDLE_ENFORCE(ctx->HasOutput(framework::GradVarName("Emission")),
+                   "Output(Emission@GRAD) should be not null.");
+    PADDLE_ENFORCE(ctx->HasOutput(framework::GradVarName("Transition")),
+                   "Output(Transition@GRAD) should be not null.");
```
Comment (qingqing01) [inline paddle/operators/linear_chain_crf_op.cc]: Remove these two checks.  The input gradients does not need to calculate at any time. Just like the old framework, some times, one branch does not need to do backward.

File: paddle/operators/linear_chain_crf_op.cc [hunk 8]
```diff
@@ -0,0 +1,570 @@
...
+                   "The Input(Label) should be a 2-D tensor with the 2nd "
+                   "dimensions fixed to 1.");
+    PADDLE_ENFORCE_EQ(
+        emission_exps_dims[0], label_dims[0],
+        "The height of Input(EmissionExps) and the height of Input(Label) "
+        "should be the same.");
+
+    ctx->SetOutputDim(framework::GradVarName("Emission"), emission_exps_dims);
+    ctx->SetOutputDim(framework::GradVarName("Transition"),
+                      transition_exps_dims);
```
Comment (qingqing01) [inline paddle/operators/linear_chain_crf_op.cc]:
```cpp
if (ctx->HasOutput(framework::GradVarName("Emission")) {
   ...
}
if (ctx->HasOutput(framework::GradVarName("Transition")) {
   ...
}
```

see `mul_op.cc`

File: paddle/operators/linear_chain_crf_op.cc [hunk 9]
```diff
@@ -0,0 +1,570 @@
...
+    }
+  }
+
+ protected:
+  void BackwardOneSequence(const platform::DeviceContext& ctx, const T ll_grad,
+                           const Tensor* emission_exps,
+                           const Tensor* transition_exps, const Tensor* alpha,
+                           const Tensor* label, Tensor* beta,
+                           Tensor* transition_grad,
+                           Tensor* emission_grad) const {
```
Comment (qingqing01) [inline paddle/operators/linear_chain_crf_op.cc]: For the const input arguments: `const Tensor&`, 
For the output arguments: `Tensor*`

File: paddle/operators/linear_chain_crf_op.h [hunk 10]
```diff
@@ -0,0 +1,54 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserve.
```
Comment (qingqing01) [inline paddle/operators/linear_chain_crf_op.h]: Will this header file also be used for the CRF decoding code? If not, it seems this header file can be removed in this PR.

File: paddle/operators/linear_chain_crf_op.cc (hunk 9 above)
Comment (lcy-seso) [reply paddle/operators/linear_chain_crf_op.cc]: Done. Thank you for pointing this out.

File: paddle/operators/linear_chain_crf_op.cc (hunk 1 above)
Comment (lcy-seso) [reply paddle/operators/linear_chain_crf_op.cc]: using auto* is suggested by our style.

File: paddle/operators/linear_chain_crf_op.cc (hunk 2 above)
Comment (lcy-seso) [reply paddle/operators/linear_chain_crf_op.cc]: auto* is suggested by our style, so I do not change this.

File: paddle/operators/linear_chain_crf_op.h (hunk 10 above)
Comment (lcy-seso) [reply paddle/operators/linear_chain_crf_op.h]: After our discussion. I will keep this header and move the function `Compute` into the header file and add GPU kernel function.

File: paddle/operators/linear_chain_crf_op.cc [hunk 11]
```diff
@@ -0,0 +1,567 @@
...
+    for (size_t i = 0; i < tag_num; ++i) {
+      alpha_value[i] = w_exps[i] * x_exps[i];
+    }
+    T ll = -x_row_max[0] - std::log(NormalizeL1<T>(alpha_value, tag_num));
+
+    for (size_t k = 1; k < seq_length; ++k) {
+      for (size_t i = 0; i < tag_num; ++i) {
+        T sum = 0.;
+        for (size_t j = 0; j < tag_num; ++j) {
+          sum += alpha_value[(k - 1) * tag_num + j] *
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply paddle/operators/linear_chain_crf_op.cc]: Done.

File: paddle/operators/linear_chain_crf_op.cc [hunk 12]
```diff
@@ -0,0 +1,567 @@
...
+    const T* w = trans_weights->data<T>();
+    const T* w_exps = trans_weight_exps->data<T>();
+    T* alpha_value = alpha->data<T>();
+
+    auto x_dims = emission->dims();
+    const size_t seq_length = x_dims[0];
+    const size_t tag_num = x_dims[1];
+    // The 1st row of w are transition weights for start mask.
+    // The 2nd row of w are transition weights for end mask.
+    // Transition weights among other tags begin from the 3rd row of w.
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply paddle/operators/linear_chain_crf_op.cc]: Done.

File: paddle/operators/linear_chain_crf_op.cc [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+
+Finaly, the linear chain CRF operator outputs the logarithm of the conditional
+likelihood of each training sample in a mini-batch.
+
+NOTE:
+1. The feature function for a CRF is made up of the emission features and the
+transition features. The emission feature weights are NOT computed in
+this operator. They MUST be computed first before this operator is called.
+
+2. Because this operator performs globally normaliztion over all possible
```
Comment (abhinavarora) [inline paddle/operators/linear_chain_crf_op.cc]: Fix typo: globally normaliztion -> *global normalization*

File: paddle/operators/linear_chain_crf_op.cc [hunk 14]
```diff
@@ -0,0 +1,260 @@
...
+                   "dimensions fixed to 1.");
+    PADDLE_ENFORCE_EQ(
+        emission_dims[0], label_dims[0],
+        "The height of Input(Emission) and the height of Input(Label) "
+        "should be the same.");
+
+    ctx->SetOutputDim("Alpha", emission_dims);
+    ctx->SetOutputDim("EmissionExps", emission_dims);
+    ctx->SetOutputDim("TransitionExps", transition_dims);
+    // (TODO caoying) This is tricky. The 1st dimension of Output(LogLikelihood)
```
Comment (abhinavarora) [inline paddle/operators/linear_chain_crf_op.cc]: You can follow the Google C++ stye guide for TODO comments. http://google.github.io/styleguide/cppguide.html#TODO_Comments

File: paddle/operators/linear_chain_crf_op.cc [hunk 15]
```diff
@@ -0,0 +1,260 @@
...
+    }
+    if (ctx->HasOutput(framework::GradVarName("Transition"))) {
+      ctx->SetOutputDim(framework::GradVarName("Transition"),
+                        transition_exps_dims);
+    }
+  }
+
+ protected:
+  // Explicitly set that the data type of output of the linear_chain_crf_grad
+  // operator is determined by its input: graidents of LogLikelihood.
```
Comment (abhinavarora) [inline paddle/operators/linear_chain_crf_op.cc]: Typo: graidents -> *gradients*

File: paddle/operators/linear_chain_crf_op.h [hunk 16]
```diff
@@ -0,0 +1,331 @@
...
+  PADDLE_ENFORCE(sum,
+                 "The unnormalized probabilities of all possible unfinished "
+                 "sequences must be greater than 0.");
+  T s = 1. / sum;
+  for (size_t i = 0; i < len; ++i) x[i] *= s;
+  return sum;
+}
+}  // namespace
+
+using framework::LoDTensor;
```
Comment (abhinavarora) [inline paddle/operators/linear_chain_crf_op.h]: This violates the Google Style Guide. For more details check out https://google.github.io/styleguide/cppguide.html#Namespaces

File: paddle/operators/linear_chain_crf_op.h [hunk 17]
```diff
@@ -0,0 +1,331 @@
...
+    auto* emission_weights = ctx.Input<LoDTensor>("Emission");
+    auto* transition_weights = ctx.Input<Tensor>("Transition");
+    auto* emission_exps = ctx.Output<LoDTensor>("EmissionExps");
+    emission_exps->mutable_data<T>(ctx.GetPlace());
+    auto* transition_exps = ctx.Output<Tensor>("TransitionExps");
+    transition_exps->mutable_data<T>(ctx.GetPlace());
+    auto* label = ctx.Input<LoDTensor>("Label");
+
+    auto in_lod = emission_weights->lod();
+    PADDLE_ENFORCE(in_lod.size(), "Input(Emission) is not a sequence.");
```
Comment (abhinavarora) [inline paddle/operators/linear_chain_crf_op.h]: Error message language could be *Input(Emission) must be a sequence*

File: paddle/operators/linear_chain_crf_op.h [hunk 18]
```diff
@@ -0,0 +1,331 @@
...
+
+    // TODO(caoying) The checks related to LoD information should be
```
Comment (abhinavarora) [inline paddle/operators/linear_chain_crf_op.h]: You can follow the Google C++ stye guide for TODO comments. http://google.github.io/styleguide/cppguide.html#TODO_Comments

File: paddle/operators/linear_chain_crf_op.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (lcy-seso) [reply paddle/operators/linear_chain_crf_op.cc]: You are quite right. It is this paper. I will add it. Thank you very much for all your comments.

**Comment by lcy-seso** [comment]
I have not finished yet, so, please do not begin the review. Thank you all.

File: paddle/framework/tensor_impl.h [hunk 19]
```diff
@@ -225,10 +225,12 @@ inline void Tensor::CopyFromVector(const std::vector<T>& src,
...
-  PADDLE_ENFORCE_GE(begin_idx, 0, "Slice begin index is less than zero.");
-  PADDLE_ENFORCE_LE(end_idx, dims_[0], "Slice end index is out of bound.");
-  PADDLE_ENFORCE_LT(begin_idx, end_idx,
-                    "Begin index must be less than end index.");
+  PADDLE_ENFORCE_GE(begin_idx, 0,
+                    "The start row index must be greater than 0.");
+  PADDLE_ENFORCE_LE(end_idx, dims_[0], "The end row index is out of bound.");
+  PADDLE_ENFORCE_LT(
+      begin_idx, end_idx,
+      "The start row index must be smaller than the end row index.");
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply paddle/framework/tensor_impl.h]: Done.

File: paddle/operators/linear_chain_crf_op.cc [hunk 20]
```diff
@@ -0,0 +1,260 @@
...
+class LinearChainCRFOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  LinearChainCRFOpMaker(framework::OpProto* proto,
+                        framework::OpAttrChecker* op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput(
+        "Emission",
+        "(LoDTensor, default: LoDTensor<float>). "
+        "The unscaled emission weight matrix for the linear chain CRF. "
+        "This input is a LoDTensor with shape [N x D] where N is the total "
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply paddle/operators/linear_chain_crf_op.cc]: Yes, I think this is no problem. Thank you. Done.

File: paddle/operators/linear_chain_crf_op.cc [hunk 21]
```diff
@@ -0,0 +1,260 @@
...
+        "element number of all input squences in a mini-batch, "
+        "and D is the total tag number.");
+    AddInput(
+        "Transition",
+        "(Tensor, default: Tensor<float>). A Tensor with shape [(D + 2) x D]. "
+        "The learnable parameter for the linear_chain_crf operator. "
+        "See more details in the operator's comments.");
+    AddInput(
+        "Label",
+        "(LoDTensor, default: LoDTensor<int>). The groundtruth which is a 2-D "
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply paddle/operators/linear_chain_crf_op.cc]: Done. Thank you.

File: paddle/operators/linear_chain_crf_op.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (lcy-seso) [reply paddle/operators/linear_chain_crf_op.cc]: Done. I add this into the operator's comments.

File: paddle/operators/linear_chain_crf_op.cc [hunk 22]
```diff
@@ -0,0 +1,260 @@
...
+    AddComment(R"DOC(
+Conditional Random Field defines an undirected probabilistic graph with nodes
+denoting random variables and edges denoting dependencies between these
+variables. CRF learns the conditional probability \f$P(Y|X)\f$, where
+\f$X = (x_1, x_2, ... , x_n)\f$ are structured inputs and
+\f$Y = (y_1, y_2, ... , y_n)\f$ are labels for the inputs.
+
+Linear chain CRF is a special case of CRF that is useful for sequence labeling
+task. Sequence labeling tasks do not assume a lot of conditional
+independences among inputs. They only concern about the input and the output
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply paddle/operators/linear_chain_crf_op.cc]: Thank you. Done.

File: paddle/operators/linear_chain_crf_op.cc [hunk 23]
```diff
@@ -0,0 +1,260 @@
...
+being linear sequences. Thus, the graph model of such a CRF is a simple chain
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply paddle/operators/linear_chain_crf_op.cc]: Done.

File: paddle/operators/linear_chain_crf_op.cc (hunk 13 above)
Comment (lcy-seso) [reply paddle/operators/linear_chain_crf_op.cc]: Done. Thank you!

File: paddle/operators/linear_chain_crf_op.cc (hunk 14 above)
Comment (lcy-seso) [reply paddle/operators/linear_chain_crf_op.cc]: Done, thank you.

File: paddle/operators/linear_chain_crf_op.cc (hunk 15 above)
Comment (lcy-seso) [reply paddle/operators/linear_chain_crf_op.cc]: Done. Thank you.

File: paddle/operators/linear_chain_crf_op.h (hunk 10 above)
Comment (lcy-seso) [reply paddle/operators/linear_chain_crf_op.h]: Done.

File: python/paddle/v2/framework/tests/test_linear_chain_crf_op.py (hunk 3 above)
Comment (lcy-seso) [reply python/paddle/v2/framework/tests/test_linear_chain_crf_op.py]: Thank you. I will fix the unit test in the next PR, for the unit test in the latest develop branch framework seems to have a certain problem.

**Comment by lcy-seso** [comment]
Some information for all the reviewers:

1. After discussion with @qingqing01, to make this operator can also run on GPU device, in this PR, I register the GPU kernel which simply copies the inputs and outputs between the CPU/GPU devices. This copy introduces many local variables and makes the codes looks not very clean. I will keep on optimizing in the following PR.
2. Unfortunately, after I merged the lasted develop branch, the previously succeeded gradient check fails.  Many related issues have been created and some the operators have been removed from the unittest (for example, the gru_unit_op, recurrent_op, lstm_unit_op, softmax_with_cross_entropy ). It seems that some changes of the gradient check framework make the unit test of this operator fail.

File: paddle/operators/linear_chain_crf_op.cc [hunk 24]
```diff
@@ -0,0 +1,569 @@
...
+    const size_t batch_size = emission_dims[0];
+    const size_t tag_num = emission_dims[1];
+    const size_t seq_num = in_lod[level].size() - 1;
+
+    Tensor emission_row_max;
+    emission_row_max.mutable_data<T>(
+        framework::make_ddim({static_cast<int>(batch_size), 1}),
+        platform::CPUPlace());
+
+    auto place = ctx.GetEigenDevice<platform::CPUPlace>();
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply paddle/operators/linear_chain_crf_op.cc]: Done.

File: paddle/operators/linear_chain_crf_op.h (hunk 17 above)
Comment (lcy-seso) [reply paddle/operators/linear_chain_crf_op.h]: Done.

File: paddle/operators/linear_chain_crf_op.h [hunk 25]
```diff
@@ -0,0 +1,543 @@
...
+    Tensor transition_exps_tensor;
+    Tensor* alpha = nullptr;
+    Tensor alpha_tensor;
+    Tensor* ll = nullptr;
+    Tensor ll_tensor;
+
+    if (platform::is_gpu_place(ctx.GetPlace())) {
+      emission_weights = &emission_weight_tensor;
+      transition_weights = &transition_weight_tensor;
+      label = &label_tensor;
```
Comment (qingqing01) [inline paddle/operators/linear_chain_crf_op.h]: For short lines code, the `LoDTensor xx` and `Tensor xx` from line 72 to line 86 can be removed, then use `new LoDTensor()` and `new Tensor()` here for the pointers from line 72 to line 86. And delete these pointers at last.

File: paddle/operators/linear_chain_crf_op.h [hunk 26]
```diff
@@ -0,0 +1,543 @@
...
+    ll->mutable_data<T>(platform::CPUPlace());
+
+    // Now, all the inputs and outputs should be on the CPU memory.
+    auto emission_dims = emission_weights->dims();
+    const size_t batch_size = emission_dims[0];
+    const size_t tag_num = emission_dims[1];
+
+    Tensor emission_row_max;
+    emission_row_max.mutable_data<T>(
+        framework::make_ddim({static_cast<int>(batch_size), 1}),
```
Comment (qingqing01) [inline paddle/operators/linear_chain_crf_op.h]:
```cpp
static_cast<int64_t>
```
The number type in `DDim` is `int64_t`.

File: paddle/operators/cross_entropy_op.cc [hunk 27]
```diff
@@ -38,8 +39,8 @@ class CrossEntropyOp : public framework::OperatorWithKernel {
                         "If Attr(soft_label) == true, the 2nd dimension of "
                         "Input(X) and Input(Label) should be equal.");
     } else {
-      PADDLE_ENFORCE_EQ(label_dims[1], 1,
-                        "If Attr(soft_label) == false, the 2nd dimension of "
+      PADDLE_ENFORCE_EQ(label_dims[1], 1UL,
```
Comment (qingqing01) [inline paddle/operators/cross_entropy_op.cc]: The dimension type is `int64_t`.

**Comment by lcy-seso** [comment]
I will enhance the unittest, and refine the codes in the following PR. Thanks for all the comments.
